File: mcp_server_development/python/src/session04/smoke.py
"""チーム稼働ダッシュボードの疎通確認クライアント（Python）

  docker compose exec python python src/session04/smoke.py
  docker compose exec python python src/session04/smoke.py src/session04/q1_server.py

このスクリプトは「クライアント側」なので print を使ってかまいません。
通信路として stdout を使っているのはサーバープロセスのほうだけで、
クライアントの stdout は人間が読む画面にすぎないからです。
"""

import asyncio
import sys

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from mcp.shared.exceptions import MCPError
from mcp.types import CallToolResult


def text_of(result: CallToolResult) -> str:
    first = result.content[0] if result.content else None
    return getattr(first, "text", "(テキストなし)")


async def main() -> None:
    entry = sys.argv[1] if len(sys.argv) > 1 else "src/session04/server.py"
    # クライアントがサーバーを子プロセスとして起動する。これが stdio トランスポートの実像
    params = StdioServerParameters(command="python", args=[entry])

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            init = await session.initialize()
            # mcp 2.0 のモデルはフィールド名が snake_case（1.x は serverInfo）
            print(f"[1/6] 接続成功: {init.server_info.name} v{init.server_info.version}")

            tools = await session.list_tools()
            names = [tool.name for tool in tools.tools]
            print(f"[2/6] tools/list: {', '.join(names)}")

            print("[3/6] list_members(team=platform):")
            print(text_of(await session.call_tool("list_members", {"team": "platform"})))

            print("[4/6] summarize_hours(2026-08-03 〜 2026-08-07):")
            print(
                text_of(
                    await session.call_tool(
                        "summarize_hours",
                        {"start_date": "2026-08-03", "end_date": "2026-08-07"},
                    )
                )
            )

            # 業務エラー：形式は正しいが期間が逆順。ツール結果として失敗が返る
            reversed_range = await session.call_tool(
                "summarize_hours",
                {"start_date": "2026-08-07", "end_date": "2026-08-03"},
            )
            body = text_of(reversed_range)
            # 文面そのものではなくフラグとキーワードで判定する（理由は本文参照）
            print(
                f"[5/6] 期間が逆順: is_error={reversed_range.is_error} / "
                f"「以前の日付」を含む: {'以前の日付' in body}"
            )

            # スキーマ違反：日付の形式が違う。どの層のエラーになるかを確認する
            try:
                broken = await session.call_tool(
                    "summarize_hours",
                    {"start_date": "2026/08/03", "end_date": "2026-08-07"},
                )
                print(f"[6/6] 日付形式違反: ツール結果で返りました（is_error={broken.is_error}）")
            except MCPError as error:
                print(f"[6/6] 日付形式違反: JSON-RPC エラー code={error.error.code}")

    print("OK: セッション4 のサーバーは TypeScript 版と同じテキストを返しています")


if __name__ == "__main__":
    asyncio.run(main())

File: mcp_server_development/node/src/session03/smoke.ts
/**
 * チーム稼働ダッシュボードの疎通確認クライアント
 *
 *   docker compose exec node npx tsx src/session03/smoke.ts
 *   docker compose exec node npx tsx src/session03/smoke.ts src/session03/lowlevel-server.ts
 *
 * このスクリプトは「クライアント側」なので console.log を使ってかまいません。
 * 通信路として stdout を使っているのはサーバープロセスのほうだけで、
 * クライアントの stdout は人間が読む画面にすぎないからです。
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";

type ToolResult = {
  isError?: boolean;
  content: Array<{ type: string; text?: string }>;
  [key: string]: unknown;
};

function textOf(result: unknown): string {
  const typed = result as ToolResult;
  const body = typed.content[0]?.text ?? "(テキストなし)";
  return typed.isError === true ? `【ツールエラー】${body}` : body;
}

const entry = process.argv[2] ?? "src/session03/server.ts";

// クライアントがサーバーを子プロセスとして起動する。これが stdio トランスポートの実像
const transport = new StdioClientTransport({ command: "npx", args: ["tsx", entry] });
const client = new Client({ name: "dashboard-smoke", version: "1.0.0" });
await client.connect(transport);

const info = client.getServerVersion();
console.log(`[1/5] 接続成功: ${info?.name} v${info?.version}`);

const { tools } = await client.listTools();
const names = tools.map((tool) => tool.name);
console.log(`[2/5] tools/list: ${names.join(", ")}`);

console.log("[3/5] list_members(team=platform):");
console.log(textOf(await client.callTool({ name: "list_members", arguments: { team: "platform" } })));

if (names.includes("summarize_hours")) {
  console.log("[4/5] summarize_hours(2026-08-03 〜 2026-08-07):");
  console.log(
    textOf(
      await client.callTool({
        name: "summarize_hours",
        arguments: { from: "2026-08-03", to: "2026-08-07" },
      }),
    ),
  );
} else {
  console.log("[4/5] summarize_hours は未登録のためスキップします");
}

// 列挙に無い値を渡す。Zod が弾き、ツール結果に isError: true が立つ
const invalid = await client.callTool({
  name: "list_members",
  arguments: { team: "unknown-team" },
});
console.log(`[5/5] 不正な引数は isError=${invalid.isError} で差し戻されました`);
console.log(`      本文: ${textOf(invalid)}`);

await client.close();
console.log("OK: セッション3 のサーバーは正常に動作しています");
